### PR TITLE
Progress bars for deploy and publish

### PR DESCRIPTION
**Pull request by trieloff**
Replaces chatty log messages with a simple progress bar:

screen recording 2018-10-18 at 16 34 59 2018-10-18 16_39_04

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
     "@adobe/parcel-plugin-htl": "^0.9.4",
     "@adobe/petridish": "^1.6.0",
     "archiver": "^3.0.0",
+    "bluebird": "^3.5.2",
```
Comment (tripodsan) [inline package.json]: noooo :-)

File: src/deploy.cmd.js [hunk 2]
```diff
@@ -293,36 +296,45 @@ class DeployCommand {
...
-            openwhisk.actions.update(actionoptions).then((result) => {
-              console.log(`✅  Action ${result.name} has been created.`);
-            }).catch((e) => {
-              console.error(`❌  Unable to deploy the action ${name}:  ${e.message}`);
-            });
-          }
-        } else {
-          console.error(`❌  File ${script} could not be read. ${err.message}`);
-        }
+    const named = scripts.map(script => ({
```
Comment (tripodsan) [inline src/deploy.cmd.js]: `named`?

File: src/deploy.cmd.js [hunk 3]
```diff
@@ -293,36 +296,45 @@ class DeployCommand {
...
+      script,
+      name: this.actionName(script),
+    }));
+
+    const read = Promise.map(named, ({ script, name }) => fs.readFile(script, { encoding: 'utf8' }).then(action => ({ script, name, action })));
```
Comment (tripodsan) [inline src/deploy.cmd.js]: better w/o bluebird:

```
read = Promise.all(named.map( { script, name} ) => .... 
```

File: package.json (hunk 1 above)
Comment (trieloff) [reply package.json]: Bluebird is already in the dependency chain, I've just moved it to top-level.

**Comment by trieloff** [comment]
It's only logging errors to the console. Do you want logging handled ad-hoc in this PR or consistently in #79 ?

In a non-TTY environment, `progress` deactivates automatically: https://github.com/visionmedia/node-progress/blob/e74b8800e0ae17486fe29e3a24dc47dee2aaf8b8/lib/node-progress.js#L120

**Comment by tripodsan** [comment]
> It's only logging errors to the console. Do you want logging handled ad-hoc in this PR or consistently in #79 ?

#79 would be nicer - but for now, I guess, a `--no-progress` bar should cause normal console logging. 

> In a non-TTY environment, progress deactivates automatically: > https://github.com/visionmedia/node-progress/blob/e74b8800e0ae17486fe29e3a24dc47dee2aaf8b8/lib/node-progress.js#L120

nice. but then it should enable normale console logging

File: package.json (hunk 1 above)
Comment (stefan-guggisberg) [reply package.json]: > Bluebird is already in the dependency chain, I've just moved it to top-level.

```
$ npm ls bluebird
@adobe/helix-cli@0.8.1-pre.14 /Users/stefan/projects/helix-cli
├─┬ @adobe/hypermedia-pipeline@0.6.0
│ ├── bluebird@3.5.2
│ └─┬ tmp-promise@1.0.5
│   └── bluebird@3.5.2  deduped
└─┬ request-promise@4.2.2
  └── bluebird@3.5.2  deduped
```

`request-promise-native` can be simply replaced with `request-promise-native`.

`@adobe/hypermedia-pipeline` would need to be ported, replacing `tmp-promise` with a module that doesn't include a bluebird dependency and removing the direct `bluebird` dependency. 

i am +1 for using native Promise unless there's a really good reason for using a Promise library..

Comment (tripodsan) [reply package.json]: me, too :-) especially since the bluebird magic in hypermedia pipeline is hard to understand :-)

**Comment by trieloff** [comment]
@tripodsan 
- bluebird is gone
- you can set `--log-file` and `--log-level` for all commands now
- when `stderr` is not a TTY, the progress bar is disabled, and status updates related to progress bar updates are logged to `stderr` (this is using the new `maybe` log level)
- when any other log *file* is set up, all messages go there
- diff coverage is lower because error code isn't tested
- I will open a separate PR for #79 in Petridish, so that this can be merged independently, for now

File: src/build.cmd.js [hunk 4]
```diff
@@ -57,8 +56,9 @@ function findStaticFiles(bnd) {
 }
 
 class BuildCommand extends EventEmitter {
-  constructor() {
+  constructor(logger = makeLogger()) {
     super();
+    this._logger = logger;
```
Comment (tripodsan) [inline src/build.cmd.js]: why not call it `this.log` ? would be a bit shorter.

File: src/clean.cmd.js [hunk 5]
```diff
@@ -54,8 +54,10 @@ class CleanCommand {
       this._cacheDir = path.resolve(this._cwd, '.hlx', 'cache');
     }
 
-    await clean(this._target);
-    await clean(this._cacheDir);
+    await Promise.all([
```
Comment (tripodsan) [inline src/clean.cmd.js]: why this change? unrelated to logger :-)

File: src/demo.cmd.js [hunk 6]
```diff
@@ -78,7 +78,7 @@ class InitCommand {
       // https://github.com/adobe/helix-cli/issues/280
       // bypass pre-commit and commit-msg hooks when doing initial commit (-n,--no-verify)
       await execAsync('git commit -q -n -m"Initial commit."');
-      this.msg(chalk.yellow('initializing git repository'));
+      this._logger.info(chalk.yellow('initializing git repository'));
```
Comment (tripodsan) [inline src/demo.cmd.js]: keep `this.msg()` as it strips out the colors for padding.

File: src/deploy.cmd.js [hunk 7]
```diff
@@ -235,12 +236,12 @@ class DeployCommand {
     await Promise.all(envars);
 
     if (follow.first_build) {
-      console.log('\nAuto-deployment started.');
```
Comment (tripodsan) [inline src/deploy.cmd.js]: maybe adding a `const log = this._logger` would help clean up the code a bit.

File: src/deploy.cmd.js (hunk 2 above)
Comment (tripodsan) [reply src/deploy.cmd.js]: can we call this `actions` instead?

File: src/log-common.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+const winston = require('winston');
+const { MESSAGE, LEVEL } = require('triple-beam');
+
+const { format } = winston;
+
+module.exports.logArgs = yargs => yargs
```
Comment (tripodsan) [inline src/log-common.js]: please use function (avoid anonymous)

File: src/perf.cmd.js [hunk 9]
```diff
@@ -133,21 +133,21 @@ class PerfCommand {
    * @param {*} limit
    * @returns true if successful, false if unsuccessful and undefined if the name isn't valid
    */
-  static format(metrics, name, limit) {
+  static format(metrics, name, limit, logger = makeLogger()) {
```
Comment (tripodsan) [inline src/perf.cmd.js]: do you need to keep this as static? consider turning into instance method.

File: src/perf.cmd.js [hunk 10]
```diff
@@ -133,21 +133,21 @@ class PerfCommand {
...
-      console.log(`  ${chalk.gray(`${metric.label}: `)}${PerfCommand.formatMeasure(metric.value, limit)}`);
+      logger.info(`  ${chalk.gray(`${metric.label}: `)}${PerfCommand.formatMeasure(metric.value, limit)}`);
       return PerfCommand.formatMeasure(metric.value, limit).indexOf('(failed)') === -1;
     }
     return undefined;
   }
 
-  static formatResponse(response, params = {}, strainname = 'default') {
-    console.log(`\nTesting ${response.url} on ${response.device.title} (${response.connection.title}) from ${response.location.emoji}  ${response.location.name} using ${strainname} strain.\n`);
+  static formatResponse(response, params = {}, strainname = 'default', logger = makeLogger()) {
```
Comment (tripodsan) [inline src/perf.cmd.js]: do you need to keep this as static? consider turning into instance method.

File: src/perf.cmd.js [hunk 11]
```diff
@@ -166,7 +166,7 @@ class PerfCommand {
   }
 
   async run() {
-    console.log(chalk.green('Testing performance…'));
```
Comment (tripodsan) [inline src/perf.cmd.js]: suggest `const log = this._logger;` for cleaner code.

File: src/deploy.cmd.js [hunk 12]
```diff
@@ -283,6 +284,13 @@ class DeployCommand {
 
     const scripts = [path.resolve(__dirname, 'openwhisk', 'static.js'), ...glob.sync(`${this._target}/*.js`)];
 
+    const bar = new ProgressBar('deploying [:bar] :etas', { total: (scripts.length * 2), width: 50, renderThrottle: 1 });
```
Comment (tripodsan) [inline src/deploy.cmd.js]: maybe expand to multiple lines...

**Comment by tripodsan** [comment]
@trieloff 

> when stderr is not a TTY, 

why only stderr?  now I have to remember `2>&1` :-)

but even then, I only get:

```
$ hlx publish --dry-run --log-level=debug 2>&1 | cat
maybe: Cloning Service Config Version 10
maybe: Skipping 3 existing backends
maybe: Skipping 14 existing dictionaries
maybe: 🌳  Set static repo
maybe: 🌳  Set content repo
maybe: Setting OpenWhisk Authentication
maybe: 🏢  Set content owner
maybe: 🏢  Set static owner
maybe: 👾  Set action root
maybe: Setting OpenWhisk namespace
maybe: ⚽️  Disable magic
maybe: Uploading VCL strains.vcl
maybe: Uploading VCL params.vcl
maybe: Uploading VCL dynamic.vcl
maybe: Uploading VCL helix.vcl
maybe: Uploaded VCL helix.vcl
maybe: Activated version 10
maybe: Purging entire cache

✅  All strains have been published and version 10 is now online.
```

not very attractive :-(

File: src/clean.cmd.js (hunk 5 above)
Comment (trieloff) [reply src/clean.cmd.js]: Yes, it's unrelated, but: `await await` enforces the sequence, `Promise.all` starts the processes in parallel and allows completion in any sequence.

File: src/perf.cmd.js (hunk 10 above)
Comment (trieloff) [reply src/perf.cmd.js]: Keeping it static makes testing easier.
